File: lib/core/utils/time_utils.dart
import 'package:flutter/material.dart';

extension TimeOfDayExtension on TimeOfDay {
  TimeOfDay addMinutes(int minutes) {
    final totalMinutes = hour * 60 + minute + minutes;
    final normalizedMinutes = totalMinutes % (24 * 60);
    return TimeOfDay(
      hour: normalizedMinutes ~/ 60,
      minute: normalizedMinutes % 60,
    );
  }

  String to24hString() {
    final hourStr = hour.toString().padLeft(2, '0');
    final minuteStr = minute.toString().padLeft(2, '0');
    return '$hourStr:$minuteStr';
  }
}

File: kmp/shared/src/commonMain/kotlin/com/jiyeong/supplementroutine/shared/scheduling/TimeUtils.kt
package com.jiyeong.supplementroutine.shared.scheduling

import com.jiyeong.supplementroutine.shared.domain.LocalDateValue
import com.jiyeong.supplementroutine.shared.domain.TimeOfDayValue

private const val MinutesPerDay = 24 * 60

fun TimeOfDayValue.addMinutes(minutes: Int): TimeOfDayValue {
    val totalMinutes = hour * 60 + minute + minutes
    val normalizedMinutes = totalMinutes.floorMod(MinutesPerDay)

    return TimeOfDayValue(
        hour = normalizedMinutes / 60,
        minute = normalizedMinutes % 60,
    )
}

fun TimeOfDayValue.to24hString(): String {
    return "${hour.twoDigits()}:${minute.twoDigits()}"
}

internal fun TimeOfDayValue.totalMinutes(): Int {
    return hour * 60 + minute
}

internal fun LocalDateValue.toRecordDateKey(): String {
    return "${year.toString().padStart(4, '0')}${month.twoDigits()}${day.twoDigits()}"
}

private fun Int.twoDigits(): String {
    return toString().padStart(2, '0')
}

private fun Int.floorMod(other: Int): Int {
    return ((this % other) + other) % other
}
